src/test/threadpool_tests.cpp

```diff
@@ -0,0 +1,267 @@
+
+    // Fill all workers with blocking tasks
+    std::vector<std::future<void>> blocking_tasks;
+    for (int i = 0; i < num_of_threads_to_block; i++) {
+        std::promise<void>& ready = ready_promises[i];
+        blocking_tasks.emplace_back(threadPool.Submit([blocker_future, &ready]() {
+            ready.set_value();
+            blocker_future.wait();
+        }));
+    }
+
+    // Wait until all threads are actually blocked
+    WaitFor(ready_futures, context);
+    return blocking_tasks;
+}
+
+BOOST_AUTO_TEST_CASE(threadpool_basic)
+{
+    // Test Cases
+    // 0) Submit task to a non-started pool.
+    // 1) Submit tasks and verify completion.
+    // 2) Maintain all threads busy except one.
+    // 3) Wait for work to finish.
+    // 4) Wait for result object.
+    // 5) The task throws an exception, catch must be done in the consumer side.
+    // 6) Busy workers, help them by processing tasks from outside.
+    // 7) Recursive submission of tasks.
+    // 8) Submit task when all threads are busy, stop pool and verify the task gets executed.
+
+    const int NUM_WORKERS_DEFAULT = 3;
+    const std::string POOL_NAME = "test";
+
+    // Test case 0, submit task to a non-started pool
+    {
+        ThreadPool threadPool(POOL_NAME);
+        bool err = false;
+        try {
+            threadPool.Submit([]() { return false; });
+        } catch (const std::runtime_error&) { err = true; }
+        BOOST_CHECK(err);
+    }
+
+    // Test case 1, submit tasks and verify completion.
+    {
+        int num_tasks = 50;
+
+        ThreadPool threadPool(POOL_NAME);
+        threadPool.Start(NUM_WORKERS_DEFAULT);
+        std::atomic<int> counter = 0;
+
+        // Store futures to ensure completion before checking counter.
+        std::vector<std::future<void>> futures;
+        futures.reserve(num_tasks);
+
+        for (int i = 1; i <= num_tasks; i++) {
+            futures.emplace_back(threadPool.Submit([&counter, i]() {
+                counter.fetch_add(i);
+            }));
+        }
+
+        // Wait for all tasks to finish
+        WaitFor(futures, /*context=*/"test1 task");
+        int expected_value = (num_tasks * (num_tasks + 1)) / 2; // Gauss sum.
+        BOOST_CHECK_EQUAL(counter.load(), expected_value);
+        BOOST_CHECK_EQUAL(threadPool.WorkQueueSize(), 0);
+    }
+
+    // Test case 2, maintain all threads busy except one.
+    {
+        ThreadPool threadPool(POOL_NAME);
+        threadPool.Start(NUM_WORKERS_DEFAULT);
+        // Single blocking future for all threads
+        std::promise<void> blocker;
+        std::shared_future<void> blocker_future(blocker.get_future());
+        const auto& blocking_tasks = BlockWorkers(threadPool, blocker_future, NUM_WORKERS_DEFAULT-1, /*context=*/"test2 blocking tasks enabled");
+
+        // Now execute tasks on the single available worker
+        // and check that all the tasks are executed.
+        int num_tasks = 15;
+        std::atomic<int> counter = 0;
```

Comment:
c219b93c3b043de202bdf3c65b433fd17af2da89

I was wondering if there's any way to assert that the unblocked tasks are all executing on one single remaining worker... would it be insane to use a non-atomic `int` here?

Reply:
> [c219b93](https://github.com/bitcoin/bitcoin/commit/c219b93c3b043de202bdf3c65b433fd17af2da89)
> 
> I was wondering if there's any way to assert that the unblocked tasks are all executing on one single remaining worker... would it be insane to use a non-atomic `int` here?

hmm, even if the non-atomic int works, that doesn't really guarantee that the increment was done in the same thread.
If we want to be 100% correct, we should store the ids of the threads that processed the tasks on a synchronized structure. Then check that only one id is there.